onnxruntime/core/providers/cpu/tensor/where_op.cc

```diff
@@ -0,0 +1,118 @@
+  ONNX_CPU_OPERATOR_TYPED_KERNEL(                                                  \
+      Where,                                                                       \
+      9,                                                                           \
+      type_name,                                                                   \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<type>()), \
+      Where<type>)
+
+#define WHERE_TYPED_KERNEL(type) \
+  WHERE_TYPED_KERNEL_WITH_TYPE_NAME(type, type)
+
+// start with a subset of types, enable more as needed...
+//WHERE_TYPED_KERNEL(uint8_t)
+//WHERE_TYPED_KERNEL(uint16_t)
+//WHERE_TYPED_KERNEL(uint32_t)
+//WHERE_TYPED_KERNEL(uint64_t)
+//WHERE_TYPED_KERNEL(int8_t)
+//WHERE_TYPED_KERNEL(int16_t)
+//WHERE_TYPED_KERNEL(int32_t)
+//WHERE_TYPED_KERNEL(int64_t)
+//WHERE_TYPED_KERNEL(MLFloat16)
+//WHERE_TYPED_KERNEL(BFloat16)
+WHERE_TYPED_KERNEL(float)
+//WHERE_TYPED_KERNEL(double)
+WHERE_TYPED_KERNEL_WITH_TYPE_NAME(std::string, string)
+//WHERE_TYPED_KERNEL(bool)
+
+#undef WHERE_TYPED_KERNEL_WITH_TYPE_NAME
+#undef WHERE_TYPED_KERNEL
+
+namespace {
+template <typename T>
+std::unique_ptr<Tensor> WhereSelection(bool target, const Tensor& condition_tensor, const Tensor& value_tensor,
+                                       TensorAllocator<T>& tensor_allocator) {
+  TBroadcaster<bool, T> selection_broadcaster{condition_tensor, value_tensor};
+  std::unique_ptr<Tensor> selection_tensor{
+      tensor_allocator.Allocate(selection_broadcaster.GetOutputShape())};
+  TBroadcastOutput<T> selection_broadcast_output{
+      selection_broadcaster.GetSpanSize(), *selection_tensor};
+
+  BroadcastLoop(
+      selection_broadcaster, selection_broadcast_output,
+      [target](EigenVectorMap<T> output, bool condition, ConstEigenVectorMap<T> value) {
+        if (condition == target) {
+          output = value;
+        } else {
+          output = EigenVectorMap<T>::PlainObject::Constant(value.size(), T{});
+        }
+      },
+      [target](EigenVectorMap<T> output, ConstEigenVectorMap<bool> condition, const T& value) {
+        output = (condition.array() == target)
+                     .select(value, EigenVectorMap<T>::PlainObject::Constant(condition.size(), T{}));
+      },
+      [target](EigenVectorMap<T> output, ConstEigenVectorMap<bool> condition, ConstEigenVectorMap<T> value) {
+        output = (condition.array() == target)
+                     .select(value, EigenVectorMap<T>::PlainObject::Constant(condition.size(), T{}));
+      });
+
+  return selection_tensor;
+}
+}  // namespace
+
+template <typename T>
+Status Where<T>::Compute(OpKernelContext* context) const {
+  const Tensor* const condition = context->Input<Tensor>(0);
+  const Tensor* const X = context->Input<Tensor>(1);
+  const Tensor* const Y = context->Input<Tensor>(2);
+  ORT_ENFORCE(condition && X && Y, "condition, X, and Y inputs are required!");
```

Comment:
May need to also implement the case where X and Y are not provided.

Reply:
That case is handled by the NonZero op.

Reply:
The op schema says condition, X and Y are required, so they will never be null as onnx\defs\schema.cc will validate they're provided.

---
In reply to: [252521567](https://github.com/Microsoft/onnxruntime/pull/412#discussion_r252521567) [](ancestors = 252521567)